proposals/mmmm-async-let.md

```diff
@@ -0,0 +1,733 @@
+
+The argument for the current semantics goes that since we did not await the task, we did not care about its result, or even failure to produce a result. 
+
+It could be argued however, that given a more complex function, with many branches in the code, we _do_ want to always await on child-tasks that may produce errors on every code path because those are important, even if not result producing values! Even in our simple `hello` function above we did actually want to ensure we waited on the registered, what we actually wanted to write is:
+
+```swift
+func hello(guest name: String) async -> String {
+  async let registered = register(name: name, delayInSeconds: 3)
+  // ... 
+  
+  _ = try await registered
+  // registration didn't throw, let's greet the guest!
+  
+  return "Hello \(name)!"
+}
+```
+
+For values and functions like the one above, where the value was _never_ used in the entire body of the function, the existing "*value was not used*" warnings should be able to help developers spot the issue. 
+
+However, in functions with more complex control flow, we wonder if this allowing to elide awaits is a good notion to follow or not. For example, the following snippet showcases a situation where the programmer made a mistake and forgot to `try await` on the `registered` result in one of the branches before returning:
+
+```swift
+func hello(guest name: String) async -> String {
+  async let registered = register(name: name, delayInSeconds: 3)
+  // ... 
+ 
+  if isFriday { 
+    print("It's Friday!")
+  } else {
+    _ = try await registered
+    // registration didn't throw, let's greet the guest!  
+    print("Any other day of the week.")
+  }
+ 
+  return "Hello \(name)!"
+}
+```
+
+By just looking at this code, it is not clear if the programmer _intentionally_ did not await on the registration on the `isFriday` branch, or if it is a real mistake, and the check must always throw. In other words, is this a place where everyone is let in on Fridays, but on other days only registered members are allowed on? :thinking: The code does not help us understand the real intent, and we would have to resort to code comments.
+
+It might be better if it were _enforced_ by the compiler to _always_ (unless throwing or returning) to have to await on all `async let` declarations. E.g. in the example above, we could detect that there exist branches on which the registered was not awaited on, and signal this as an error to the programmer, who would have to:
+
+- either fill in the appropriate `try await registered` inside the `isFriday` branch, or
+- move the `async let registered` declaration into the `else` branch — we indeed only perform this check on non-Fridays.
+
+This rule might be too cumbersome for some code though, so perhaps this warrants a future extension where it is possible to require `@exhaustiveSpawnLetWaiting` on function level, to enforce that async lets are awaited on all code paths.
```

Comment:
```suggestion
This rule might be too cumbersome for some code though, so perhaps this warrants a future extension where it is possible to require `@exhaustiveAsyncLetWaiting` on function level, to enforce that async lets are awaited on all code paths.
```

Reply:
thanks; eh I'm terrible at those mass renames it seems